Test failure looks real:

```
FAIL: corehq.apps.users.tests.test_location_assignment:CCUserLocationAssignmentTest.test_deleting_location_updates_user
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/users/tests/test_location_assignment.py", line 104, in test_deleting_location_updates_user
    self.assertAssignedLocations([])
  File "/mnt/commcare-hq-ro/corehq/apps/users/tests/test_location_assignment.py", line 130, in assertAssignedLocations
    self.assertListEqual(user.assigned_location_ids, expected_location_ids)
AssertionError: Lists differ: [u'5b9f665cdd9a419c9b5d6d89811... != []
First list contains 2 additional elements.
First extra element 0:
u'5b9f665cdd9a419c9b5d6d89811f45dc'
- [u'5b9f665cdd9a419c9b5d6d89811f45dc', u'68ba2c203b624a8e9bb7566925e8a040']
+ []
```

--- a/corehq/apps/locations/tests/test_location_fixtures.py
+++ b/corehq/apps/locations/tests/test_location_fixtures.py
@@ -21,6 +21,8 @@ from corehq.apps.users.models import CommCareUser
 from corehq.apps.app_manager.tests.util import TestXmlMixin, extract_xml_partial
 from corehq.apps.users.dbaccessors.all_commcare_users import delete_all_users
 
+from corehq.elastic import refresh_elasticsearch_index
+
 from .util import (
     setup_location_types_with_structure,
     setup_locations_with_structure,
@@ -740,13 +742,32 @@ class ShouldSyncLocationFixturesTest(TestCase):
             should_sync_locations(SyncLog(date=after_save), locations_queryset, self.user.to_ota_restore_user())
         )
 
+        # Set the user's location
+        self.user.set_location(location)
+        after_assign = datetime.utcnow()
+
+        # Should resync if last sync was before location was assigned
+        self.assertTrue(
+            should_sync_locations(SyncLog(date=after_save), locations_queryset, self.user.to_ota_restore_user())
+        )
+        # Should not resync if last sync was after location was assigned
+        self.assertFalse(
+            should_sync_locations(SyncLog(date=after_assign), locations_queryset, self.user.to_ota_restore_user())
+        )
+
         # Delete the location
+        refresh_elasticsearch_index('users')
         location.full_delete()
         after_delete = datetime.utcnow()
 
-        # TODO: Should resync if last sync was after location was saved but before location was deleted
-
-        # TODO: Should not resync if last sync was after location was deleted
+        # Should resync if last sync was before location was deleted
+        self.assertTrue(
+            should_sync_locations(SyncLog(date=after_assign), locations_queryset, self.user.to_ota_restore_user())
+        )
+        # Should not resync if last sync was after location was deleted
+        self.assertFalse(
+            should_sync_locations(SyncLog(date=after_delete), locations_queryset, self.user.to_ota_restore_user())
+        )
 
 
 @mock.patch('corehq.apps.domain.models.Domain.uses_locations', lambda: True)

--- a/corehq/apps/locations/tests/test_location_fixtures.py
+++ b/corehq/apps/locations/tests/test_location_fixtures.py
@@ -701,6 +701,7 @@ class ShouldSyncLocationFixturesTest(TestCase):
         after_save = datetime.utcnow()
         self.assertEqual('winterfell', location.name)
         locations_queryset = SQLLocation.objects.filter(pk=location.pk)
+        # Should not resync if last sync was after location save
         self.assertFalse(
             should_sync_locations(SyncLog(date=after_save), locations_queryset, self.user.to_ota_restore_user())
         )
@@ -711,12 +712,41 @@ class ShouldSyncLocationFixturesTest(TestCase):
 
         location = SQLLocation.objects.last()
         locations_queryset = SQLLocation.objects.filter(pk=location.pk)
+        # Should resync if last sync was after location was saved but before location was archived
         self.assertTrue(
             should_sync_locations(SyncLog(date=after_save), locations_queryset, self.user.to_ota_restore_user())
         )
+        # Should not resync if last sync was after location was deleted
         self.assertFalse(
             should_sync_locations(SyncLog(date=after_archive), locations_queryset, self.user.to_ota_restore_user())
         )
+
+    def test_deleting_location_should_resync(self):
+        """
+        When locations are deleted, we should resync them
+        """
+        location = make_location(
+            domain=self.domain,
+            name="Riverrun",
+            location_type=self.location_type.name,
+        )
+        location.save()
+        after_save = datetime.utcnow()
+        self.assertEqual('Riverrun', location.name)
+        locations_queryset = SQLLocation.objects.filter(pk=location.pk)
+
+        # Should not resync if last sync was after location was saved
+        self.assertFalse(
+            should_sync_locations(SyncLog(date=after_save), locations_queryset, self.user.to_ota_restore_user())
+        )
+
+        # Delete the location
+        location.full_delete()
+        after_delete = datetime.utcnow()
+
+        # TODO: Should resync if last sync was after location was saved but before location was deleted
+
+        # TODO: Should not resync if last sync was after location was deleted
 
 
 @mock.patch('corehq.apps.domain.models.Domain.uses_locations', lambda: True)

--- a/corehq/apps/users/tests/test_location_assignment.py
+++ b/corehq/apps/users/tests/test_location_assignment.py
@@ -6,6 +6,7 @@ from django.test import TestCase
 
 from corehq.apps.commtrack.tests.util import make_loc
 from corehq.apps.domain.shortcuts import create_domain
+from corehq.elastic import refresh_elasticsearch_index
 from corehq.apps.locations.tests.util import delete_all_locations
 from corehq.apps.users.models import CommCareUser, WebUser
 from corehq.apps.users.management.commands import add_multi_location_property
@@ -99,6 +100,7 @@ class CCUserLocationAssignmentTest(TestCase):
 
     def test_deleting_location_updates_user(self):
         self.user.reset_locations(self.loc_ids)
+        refresh_elasticsearch_index('users')
         self.loc1.sql_location.full_delete()
         self.loc2.sql_location.full_delete()
         self.assertAssignedLocations([])

